codegen/core/src/main/scala/com/lightbend/akkasls/codegen/SourceGeneratorUtils.scala

```diff
@@ -181,38 +180,61 @@ object SourceGeneratorUtils {
     else dotsToCamelCase(s.tail, resultSoFar + s.head, false)
   }
 
-  class CodeBlock(val code: Seq[Any]) {
+  sealed trait CodeElement
+  object CodeElement {
+    case class StringElement(string: String) extends CodeElement
+    case class FQNElement(fqn: FullyQualifiedName) extends CodeElement
+
+    implicit def stringElement(string: String): CodeElement = StringElement(string)
+    implicit def fqnElement(fqn: FullyQualifiedName): CodeElement = FQNElement(fqn)
+    implicit def block(els: Iterable[CodeElement]): CodeElement =
+      CodeBlock(
+        // add line terminator between each two elements
+        els.toVector.flatMap(Seq(_, stringElement("\n"))).dropRight(1))
+  }
+  case class CodeBlock(code: Seq[CodeElement]) extends CodeElement {
+    import CodeElement._
+
     code.foreach(validateType)
 
     private def validateType(seq: Any): Unit = seq match {
-      case _: String             => ()
-      case _: FullyQualifiedName => ()
-      case block: CodeBlock      => block.code.foreach(validateType)
-      case s: Seq[_]             => s.foreach(validateType)
-      case other =>
-        throw new IllegalArgumentException(s"Unexpected value of type [${other.getClass}] in block: [$other]")
+      case _: StringElement => ()
+      case _: FQNElement    => ()
+      case block: CodeBlock => block.code.foreach(validateType)
     }
 
     /** All classes used in this code block */
     def fqns: Seq[FullyQualifiedName] = code.flatMap {
-      case name: FullyQualifiedName => Seq(name)
-      case block: CodeBlock         => block.fqns
-      case seq: Seq[_]              => new CodeBlock(seq).fqns
-      case _                        => Seq.empty
+      case FQNElement(name) => Seq(name)
+      case block: CodeBlock => block.fqns
+      case _: StringElement => Seq.empty
     }
+
+    def write(imports: Imports, typeName: (FullyQualifiedName, Imports) => String): String =
```

Comment:
I guess this could even be

```suggestion
    def write(typeName: FullyQualifiedName => String): String =
```

but perhaps not worth the change

Reply:
:+1: It's worth and simple to do :)